### ISSUE TITLE
Completely removing the dependence on root

### ISSUE THREAD

**Issue by Dani13567** [chenxiaolong/DualBootPatcher#1452]
Back on Oct 28, 2015 you [mentioned](https://github.com/chenxiaolong/DualBootPatcher/commit/7414b4e1bd9375c718eaad7ecb670a9623bb9230):

> ... Dropping this feature will allow me to focus on other important issues like **completely
removing the dependence on root**.

I am thinking about continuing the development of this project to support the recent versions of Android too. One feature that is crucial for my project is to know if, theoretically, it is possible to install DualBootPatcher on non-rooted devices too. Would you please let me know if this can be done?

**Comment by klorinczi** [comment]
I would welcome it if you could continue the development of DualBootPatcher to support the recent versions of Android!

**Comment by chenxiaolong** [comment]
> I am thinking about continuing the development of this project to support the recent versions of Android too. One feature that is crucial for my project is to know if, theoretically, it is possible to install DualBootPatcher on non-rooted devices too. Would you please let me know if this can be done?

This is actually already implemented. So for example, if on your (bootloader unlocked) device, you flash a patched ROM, all of the features in the DualBootPatcher app are immediately available without flashing any root solution, like Magisk. This works because patched ROMs run a daemon very early in the boot process with root privileges that the app communicates with.

A root solution (like Magisk) is usually only required in two situations:
* You're booted into the stock/unpatched ROM and want to flash a secondary ROM from the app itself instead of using TWRP. In this case, the app uses root to start the daemon and then communicates with the daemon directly.
* You're switching from one developer's DualBootPatcher build to another developer's, which will have a different signing key. The daemon that runs at boot will only ever talk to an app signed by the same key for security reasons. Root is used to kill the existing daemon and start the new one.

So to summarize, the device doesn't have to be rooted, but it does need to have an unlocked bootloader. It will never be possible to use DualBootPatcher on a bootloader-locked device without an exploit.

> I would welcome it if you could continue the development of DualBootPatcher to support the recent versions of Android!

Thanks! I'm not ruling this out for the future, but I'm still pretty burnt out. If I'm being honest, I probably won't be working on this project again this year.
